torch/nested/_internal/ops.py

```diff
@@ -255,19 +267,461 @@ def jagged_binary_pointwise(func, *args, **kwargs):
     raise RuntimeError(mismatch_error_msg)
 
 
+def _validate_sdpa_input(
+    query: torch.Tensor,
+    key: torch.Tensor,
+    value: torch.Tensor,
+    attn_mask: Optional[torch.Tensor] = None,
+    dropout_p=0.0,
+    is_causal=False,
+    scale=None,
+):
+    if (
+        not isinstance(query, NestedTensor)
+        or not isinstance(key, NestedTensor)
+        or not isinstance(value, NestedTensor)
+    ):
+        raise ValueError(
+            f"Expected query, key, and value to be nested tensors, "
+            f"but got query.is_nested: {query.is_nested}, key.is_nested: {key.is_nested}, "
+            f"and value.is_nested: {value.is_nested} instead."
+        )
+    if query.dtype != key.dtype or query.dtype != value.dtype:
+        raise ValueError(
+            f"Expected query, key, and value to have the same dtype, "
+            f"but got query.dtype: {query.dtype}, key.dtype: {key.dtype}, "
+            f"and value.dtype: {value.dtype} instead."
+        )
+    if query.device != key.device or query.device != value.device:
+        raise ValueError(
+            f"Expected query, key, and value to have the same device type, "
+            f"but got query.device: {query.device}, key.device: {key.device}, "
+            f"and value.device: {value.device} instead."
+        )
+    if query.dim() < 2 or key.dim() < 2 or value.dim() < 2:
+        raise ValueError(
+            f"Expected query, key, and value to all be  at least 2 dimensional, but got query.dim: "
+            f"{query.dim()}, key.dim: {key.dim()} and value.dim: {value.dim()} instead."
+        )
+    if query._ragged_idx != 2 or key._ragged_idx != 2 or value._ragged_idx != 2:
+        raise ValueError(
+            f"Expected query, key, and value to all be be jagged at dimension 2, but got query._ragged_idx: "
+            f"{query._ragged_idx}, key._ragged_idx: {key._ragged_idx} and value._ragged_idx: {value._ragged_idx} instead."
+        )
```

Comment:
Note that in SAM we have a case where the ragged dim is a batch dim and the sequence length is constant: shape `(B, *, S, D)`. So this check is too restrictive in that scenario.

Now that I think about it, there may be other places within the preprocessing code that assume `ragged dim == sequence length dim`. I think the constant sequence length case is relatively easy to handle computation-wise though, and we do need it. I'll recommend a test case for this scenario in addition to testing the standard case.

Reply:
ok, I'll go through the code, as I agree most of the C++ NT SDPA code assumes raggedness on the sequence length. The issue might be that the fused NT SDPA kernels are also written with that assumption in mind. @drisspg might know better

Reply:
@drisspg said that this use case is not supported in SDPA as written right now. I guess just using nt.values() as the input might work?

Reply:
> @drisspg said that this use case is not supported in SDPA as written right now. I guess just using nt.values() as the input might work?

I'm doing this use case for SAM now without SDPA knowing :p

Shape `(B, *, S, D)` would be representable by a dense tensor of shape `(sum(*), S, D)`. SDPA accepts that and treat the `sum(*)` as just a normal batch dimension. The fact that it's "ragged" is maintained outside of SDPA.

Reply:
hmmm this probably requires rewiring the C++ NT implementation for SDPA too, as it currently will go to the ragged seq lens kernels the moment it detects any kind of nested tensor

Reply:
correct flash_forward only knows about `(sum(*), num_heads, d)`

Reply:
I've added a TODO on my list for the next PR to work out the semantics for SDPA + SAM, as it definitely requires some more work on the C++ code, which I think is out of scope for this PR

Reply:
> hmmm this probably requires rewiring the C++ NT implementation for SDPA too, as it currently will go to the ragged seq lens kernels the moment it detects any kind of nested tensor

I'm confused about this- we'd be sending it a dense tensor buffer, not a nested tensor. At least that's what I'm doing now. 

> I've added a TODO on my list for the next PR to work out the semantics for SDPA + SAM, as it definitely requires some more work on the C++ code, which I think is out of scope for this PR

Cool, I'm fine with doing this in a follow-up, but I'm also very unclear why we have to make C++ changes if I'm already handling the SAM use case from python now